src/poke_env/player/openai_api.py

```diff
@@ -0,0 +1,537 @@
+from poke_env.player.battle_order import BattleOrder, ForfeitBattleOrder
+from poke_env.player.player import Player
+from poke_env.player_configuration import PlayerConfiguration
+from poke_env.server_configuration import (
+    ServerConfiguration,
+    LocalhostServerConfiguration,
+)
+from poke_env.teambuilder.teambuilder import Teambuilder
+
+ObservationType = TypeVar("ObservationType")
+ActionType = TypeVar("ActionType")
+
+
+def __run_loop(loop: asyncio.AbstractEventLoop):
+    asyncio.set_event_loop(loop)
+    loop.run_forever()
+
+
+def __stop_loop(loop: asyncio.AbstractEventLoop, thread: Thread):  # pragma: no cover
+    disable(CRITICAL)
+    tasks = []
+    if py_ver.major == 3 and py_ver.minor >= 7:
+        caller = asyncio
+    else:
+        caller = asyncio.Task
+    for task in caller.all_tasks(loop):
+        task.cancel()
+        tasks.append(task)
+    cancelled = False
+    shutdown = asyncio.run_coroutine_threadsafe(loop.shutdown_asyncgens(), loop)
+    shutdown.result()
+    while not cancelled:
+        cancelled = True
+        for task in tasks:
+            if not task.done():
+                cancelled = False
+    loop.call_soon_threadsafe(loop.stop)
+    thread.join()
+    loop.call_soon_threadsafe(loop.close)
+
+
+def __clear_loop():  # pragma: no cover
+    __stop_loop(THREAD_LOOP, _t)
+
+
+MAIN_LOOP = asyncio.get_event_loop()
+THREAD_LOOP = asyncio.new_event_loop()
+py_ver = sys.version_info
+_t = Thread(target=__run_loop, args=(THREAD_LOOP,), daemon=True)
+_t.start()
+atexit.register(__clear_loop)
+
+
+class EnvLoop(AbstractContextManager):
+    def __enter__(self):
+        global MAIN_LOOP
+        MAIN_LOOP = asyncio.get_event_loop()
+        asyncio.set_event_loop(THREAD_LOOP)
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        asyncio.set_event_loop(MAIN_LOOP)
+
+
+class _AsyncQueue:
+    def __init__(self, queue: asyncio.Queue):
+        if not isinstance(queue, asyncio.Queue):
+            raise RuntimeError(f"Expected asyncio.Queue, got {type(queue)}")
+        self.queue = queue
+
+    async def async_get(self):
+        return await self.queue.get()
+
+    def get(self):
+        res = asyncio.run_coroutine_threadsafe(
+            self.queue.get(), asyncio.get_event_loop()
+        )
+        return res.result()
+
+    async def async_put(self, item):
+        await self.queue.put(item)
+
+    def put(self, item):
+        task = asyncio.run_coroutine_threadsafe(
+            self.queue.put(item), asyncio.get_event_loop()
+        )
+        task.result()
+
+    def empty(self):
+        return self.queue.empty()
+
+    def join(self):
+        task = asyncio.run_coroutine_threadsafe(
+            self.queue.join(), asyncio.get_event_loop()
+        )
+        task.result()
+
+    async def async_join(self):
+        await self.queue.join()
+
+
+class _AsyncPlayer(Player):
+    def __init__(self, user_funcs, username, **kwargs):
+        self.__class__.__name__ = username
```

Comment:
What is this for?

Reply:
This avoids all the players to use as username "_AsyncPlayer xyz". It instead uses the name of the class that contains the player

Reply:
It is a bit hacky, I know ;)

Reply:
Oh, got it. Definitely hacky, but that's ok :)